oshmem/include/shmem.h.in

```diff
@@ -207,6 +218,581 @@ extern shmem_ctx_t oshmem_ctx_default;
+
+
+
+/* Teams reduction: AND */
+OSHMEM_DECLSPEC  int shmem_uchar_and_reduce(shmem_team_t team, unsigned char *target, const unsigned char *source, size_t nreduce);
+OSHMEM_DECLSPEC  int shmem_ushort_and_reduce(shmem_team_t team, unsigned short *target, const unsigned short *source, size_t nreduce);
+OSHMEM_DECLSPEC  int shmem_uint_and_reduce(shmem_team_t team, unsigned int *target, const unsigned int *source, size_t nreduce);
+OSHMEM_DECLSPEC  int shmem_ulong_and_reduce(shmem_team_t team, unsigned long *target, const unsigned long *source, size_t nreduce);
+OSHMEM_DECLSPEC  int shmem_ulonglong_and_reduce(shmem_team_t team, unsigned long long *target, const unsigned long long *source, size_t nreduce);
+OSHMEM_DECLSPEC  int shmem_int_and_reduce(shmem_team_t team, int *target, const int *source, size_t nreduce);
+OSHMEM_DECLSPEC  int shmem_longlong_and_reduce(shmem_team_t team, long long *target, const long long *source, size_t nreduce);
+OSHMEM_DECLSPEC  int shmem_int8_and_reduce(shmem_team_t team, int8_t *target, const int8_t *source, size_t nreduce);
+OSHMEM_DECLSPEC  int shmem_int16_and_reduce(shmem_team_t team, int16_t *target, const int16_t *source, size_t nreduce);
+OSHMEM_DECLSPEC  int shmem_int32_and_reduce(shmem_team_t team, int32_t *target, const int32_t *source, size_t nreduce);
+OSHMEM_DECLSPEC  int shmem_int64_and_reduce(shmem_team_t team, int64_t *target, const int64_t *source, size_t nreduce);
+OSHMEM_DECLSPEC  int shmem_uint8_and_reduce(shmem_team_t team, uint8_t *target, const uint8_t *source, size_t nreduce);
+OSHMEM_DECLSPEC  int shmem_uint16_and_reduce(shmem_team_t team, uint16_t *target, const uint16_t *source, size_t nreduce);
+OSHMEM_DECLSPEC  int shmem_uint32_and_reduce(shmem_team_t team, uint32_t *target, const uint32_t *source, size_t nreduce);
+OSHMEM_DECLSPEC  int shmem_uint64_and_reduce(shmem_team_t team, uint64_t *target, const uint64_t *source, size_t nreduce);
+OSHMEM_DECLSPEC  int shmem_size_and_reduce(shmem_team_t team, size_t *target, const size_t *source, size_t nreduce);
+#if OSHMEM_HAVE_C11
+#define shmem_and_reduce(...)                                          \
+    _Generic(&*(__OSHMEM_VAR_ARG1(__VA_ARGS__)),                       \
+            shmem_team_t: _Generic(&*(__OSHMEM_VAR_ARG2(__VA_ARGS__)), \
+                unsigned char*:      shmem_uchar_and_reduce,           \
+                unsigned short*:     shmem_ushort_and_reduce,          \
+                unsigned int*:       shmem_uint_and_reduce,            \
+                unsigned long*:      shmem_ulong_and_reduce,           \
+                unsigned long long*: shmem_ulonglong_and_reduce,       \
+                int8_t*: shmem_int8_and_reduce,                        \
+                int16_t*: shmem_int16_and_reduce,                      \
+                int32_t*: shmem_int32_and_reduce,                      \
+                int64_t*: shmem_int64_and_reduce,                      \
```

Comment:
why used intXX_t datatypes instead of generic types? in some other macro used generic datatypes for intXX_t buffers

Reply:
Please refer to Table 10 of the v1.5 user guide. I was following Table 10. Later on, After further discussion, we realized that I have to support Table 11 as well. That's why there is this slight inconsistency. Do you want me to change them? or it is ok as it is?